ddtrace/context.py

```diff
@@ -123,11 +130,74 @@ def sampling_priority(self, value):
     @property
     def _traceparent(self):
         # type: () -> str
-        if self.trace_id is None or self.span_id is None:
+        tp = self._meta.get(_TRACEPARENT_KEY)
+        if self.span_id is None or self.trace_id is None:
+            if tp:
+                # if we only have a traceparent then we'll forward it
+                return tp
+            # if we don't have a span id or trace id value we can't build a valid traceparent
             return ""
 
+        # determine the trace_id value
+        if tp:
+            # grab the original traceparent trace id, not the converted value
+            trace_id = tp.split("-")[1]
+        else:
+            trace_id = "{:032x}".format(self.trace_id)
+
         sampled = 1 if self.sampling_priority and self.sampling_priority > 0 else 0
-        return "00-{:032x}-{:016x}-{:02x}".format(self.trace_id, self.span_id, sampled)
+        return "00-{}-{:016x}-{:02x}".format(trace_id, self.span_id, sampled)
+
+    @property
+    def _tracestate(self):
+        # type: () -> str
+        dd = ""
+        if self.sampling_priority:
+            dd += "s:{};".format(self.sampling_priority)
+        if self.dd_origin:
+            dd += "o:{};".format(self.dd_origin)
+        sampling_decision = self._meta.get(SAMPLING_DECISION_TRACE_TAG_KEY)
+        if sampling_decision:
+            # replace characters ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+            dd += "t.dm:{};".format(re.sub(_TRACESTATE_INVALID_CHARS_REGEX, "_", sampling_decision))
+        # since this can change, we need to grab the value off the current span
+        usr_id_key = self._meta.get(USER_ID_KEY)
+        if usr_id_key:
+            dd += "t.usr.id:{};".format(re.sub(_TRACESTATE_INVALID_CHARS_REGEX, "_", usr_id_key))
+
+        # grab all other _dd.p values out of meta since we need to propagate all of them
+        for k, v in self._meta.items():
+            if (
+                isinstance(k, six.string_types)
+                and k.startswith("_dd.p")
+                # we've already added sampling decision and user id
+                and k not in [SAMPLING_DECISION_TRACE_TAG_KEY, USER_ID_KEY]
+            ):
+                # for key replace ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+                # for value replace ",", ";", ":" and characters outside the ASCII range 0x20 to 0x7E
+                next_tag = "{}:{};".format(
+                    re.sub("_dd.p.", "t.", re.sub(r",| |=|[^\x20-\x7E]+", "_", k)),
+                    re.sub(_TRACESTATE_INVALID_CHARS_REGEX, "_", v),
+                )
+                if not (len(dd) + len(next_tag)) > 256:
+                    dd += next_tag
+        # remove final ;
+        if dd:
+            dd = dd[:-1]
```

Comment:
This could be avoided if we append to a list while going through span tags which we then join once done.